File: 3628-find-minimum-time-to-reach-last-room-ii/3628-find-minimum-time-to-reach-last-room-ii.cs
public class Solution {
    public int MinTimeToReach(int[][] moveTime) {
        int m = moveTime.Length, n = moveTime[0].Length;
        (int, int)[] dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)];

        var minTime = new int[m * n];
        var pq = new PriorityQueue<int[], int>();
        pq.Enqueue([0, 0, 0, 0], 0);  // [time, extra, x, y] => extra = 1 if next step needs 2 secs

        while (pq.Count > 0) {
            var val = pq.Dequeue();
            int tick = val[0], extra = val[1], x = val[2], y = val[3];
            var minTimeIdx = x * n + y;

            if(!(x == 0 && y == 0) && minTime[minTimeIdx] > 0) continue;    // already processed with better time
            if(x == m - 1 && y == n - 1) return tick;                       // found destination
            minTime[minTimeIdx] = tick;

            foreach(var (dx, dy) in dirs){
                int nx = x + dx, ny = y + dy;
                if(nx == -1 || ny == -1 || nx == m || ny == n) continue;    // out of bounds

                var nextTick = 1 + extra + Math.Max(tick, moveTime[nx][ny]);
                var nextIdx = nx * n + ny;
                if(minTime[nextIdx] > 0 && minTime[nextIdx] <= nextTick) continue;
                pq.Enqueue([nextTick, 1 - extra, nx, ny], nextTick);
            }
        }

        return -1;
    }
}

File: 3628-find-minimum-time-to-reach-last-room-ii/3628-find-minimum-time-to-reach-last-room-ii.py
class Solution:
    def minTimeToReach(self, moveTime: List[List[int]]) -> int:
        m, n = len(moveTime), len(moveTime[0])
        dirs = [(0,1), (0,-1), (1,0), (-1,0)]

        minTime = [0] * (m * n)
        pq = [(0, 0, 0, 0)]  # [time, extra, x, y] => extra = 1 if next step needs 2 secs

        while pq:
            tick, extra, x, y = heappop(pq)
            minTimeIdx = x * n + y

            if not (x == 0 and y == 0) and minTime[minTimeIdx] > 0: continue    # already processed with better time
            if x == m - 1 and y == n - 1: return tick                           # found destination
            minTime[minTimeIdx] = tick

            for dx, dy in dirs:
                nx, ny = x + dx, y + dy
                if nx < 0 or ny < 0 or nx >= m or ny >= n: continue             # out of bounds
                nextTick = 1 + extra + max(tick, moveTime[nx][ny])
                nextIdx = nx * n + ny
                if minTime[nextIdx] > 0 and minTime[nextIdx] <= nextTick: continue
                heappush(pq, (nextTick, 1 - extra, nx, ny))

        return -1   # unreachable